components/hab/src/command/pkg/uninstall.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs;
+use std::path::Path;
+
+use common::package_graph::PackageGraph;
+use common::ui::{Status, UIWriter, UI};
+use error::{Error, Result};
+use hcore::fs as hfs;
+use hcore::package::{all_packages, PackageIdent, PackageInstall};
+
+//use std::iter::IntoIterator;
+
+///
+/// Delete a package and all dependencies which are not used by other packages
+///
+/// 1. We find the fully qualified package ident and all its dependencies
+/// 2. We find all packages on the filesystem and convert them into a graph
+/// 3. We do a BFS on the graph to get the dependencies in order
+/// 4. We check if the specified package has any reverse deps
+///     4a. If there are, we throw an error
+///     4b. If not, we delete the package
+/// 5. For each dependency we check if there are any packages which depend on it
+///     5a. If there are, we skip it
+///     5b. If there are not, we delete it from disk and the graph
+///
+pub fn start(
+    ui: &mut UI,
+    ident: &PackageIdent,
+    fs_root_path: &Path,
+    dryrun: bool,
+    no_deps: bool,
+) -> Result<()> {
+    // 1.
+    let pkg_install = PackageInstall::load(ident, Some(fs_root_path))?;
+    let ident = pkg_install.ident();
+    ui.begin(format!("Uninstalling {}", &ident))?;
+
+    // 2.
+    let packages = load_all_packages(&fs_root_path)?;
+    let mut graph = PackageGraph::new();
+    graph.build(&packages)?;
+
+    // 3.
+    let deps = graph.ordered_deps(&ident);
+
+    // 4.
+    match graph.count_rdeps(&ident) {
+        None => {
+            // package not in graph - this shouldn't happen but could be a race condition with another hab uninstall. We can continue as
+            // what we wanted (package to be removed) has already happened
+            ui.warn(format!("Tried to find dependencies of {} but it wasn't in graph.  Maybe another uninstall command was run at the same time?", &ident))?;
```

Comment:
Might be worth saying explicitly (even if just in comments) that we're going to continue because there may be dependencies (which we've already computed at this point) that we'll need to delete, too.

Reply:
Also, wouldn't this be searching for "dependents on" rather than "dependencies of"?